### PR TITLE
First commit fo visibility, WIP fixing shortcuts.go

### PR DESCRIPTION
**Pull request by snkzt**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+// TODO: add language flag diversion, unit unit test
+
+func main() {
+	// peeper subcommand "get"
+	getCmd := flag.NewFlagSet("get", flag.ExitOnError)
+	// Flags for "get" subcommand
+	// ("flag name", default, "Flag explanation")
+	getAll := getCmd.Bool("all", false, "Get full shortcut list")
+	getByKeyword := getCmd.String("keyword", "", "Find a shortcut with keyword ")
```
Comment (kinbiko) [inline main.go]: Presumably these will have to be mutually exclusive, right? In that case, maybe one should be the default, and one should be the flag that trigger special behaviour.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+	// peeper subcommand "add"
+	addCmd := flag.NewFlagSet("add", flag.ExitOnError)
+	// Flags(inputs) for "add" subcommand
+	addName := addCmd.String("name", "", "Name of the shortcut")
+	addShortcut := addCmd.String("shortcut", "", "Shortcut")
```
Comment (kinbiko) [inline main.go]: Presumably both of these are required right? In that case, perhaps they should be arguments (instead of flags)

File: main.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		HandleGet(getCmd, getAll, getByKeyword)
+	case "add":
+		HandleAdd(addCmd, addName, addShortcut)
+	case "delete":
+		HandleDelete(deleteCmd, deleteAll, deleteByName)
+		// TODO: Do we need to add default and return error for the case of the command doesn't exist?
+	}
+}
+
+func HandleGet(getCmd *flag.FlagSet, all *bool, keyword *string) {
```
Comment (kinbiko) [inline main.go]: I don't think any of your functions need to be public (yet).

File: main.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	getCmd.Parse(os.Args[2:])
+
+	if *all == false && *keyword == "" {
```
Comment (kinbiko) [inline main.go]: Are these de-references safe, or is there a chance that this could panic? (They might be safe, I can't remember)

File: main.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		}
+	}
+}
+
+func ValidateNewShortcutKey(addCmd *flag.FlagSet, name *string, shortcut *string) {
+	addCmd.Parse(os.Args[2:])
+	if *name == "" || *shortcut == "" {
+		fmt.Print("Name and the shortcut key are required to add a shortcut key")
+		addCmd.PrintDefaults()
+		os.Exit(1)
```
Comment (kinbiko) [inline main.go]: Can you refactor your code to return errors instead, and only have a single os.Exit(1) in your `main` function?

[Here's a blog post with a pattern that I like that might help you](https://pace.dev/blog/2020/02/12/why-you-shouldnt-use-func-main-in-golang-by-mat-ryer.html), or more concretely [my own template](https://github.com/kinbiko/go-template/blob/9f171de74003e153e12ba72df93c5afc50cd20f1/cmd/app/main.go#L8-L17) with this pattern.

File: shortcuts.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	}
+}
+
+func deleteShortcut(name string) {
+	// Get existing shortcut key list
+	shortcuts := getShortcuts()
+	for _, shortcut := range shortcuts {
+		var mapStruct map[string]json.RawMessage
+		if err := json.Unmarshal([]byte(shortcut), &mapStruct); err != nil {
+			panic(err)
```
Comment (kinbiko) [inline shortcuts.go]: Boo 👎🙂 
You should basically never use panic, and this is not an exception. Return errors instead.

**Review by kinbiko — CHANGES_REQUESTED** [review]
I'm impressed -- this is pretty good stuff right away! Keep it up!

File: main.go (hunk 1 above)
Comment (snkzt) [reply main.go]: Thank you will change so!

File: main.go (hunk 2 above)
Comment (snkzt) [reply main.go]: >both of these are required

Yes as {"Quite the app", "Cmd + Q"}

> perhaps they should be arguments

Thank you, will find how to do this.

File: main.go (hunk 3 above)
Comment (snkzt) [reply main.go]: Okie!
Still not sure about scope for function and structure, so will make sure.

File: main.go (hunk 4 above)
Comment (snkzt) [reply main.go]: >de-references safe

Was not aware of this point, thanks, will update this part later.

File: main.go (hunk 5 above)
Comment (snkzt) [reply main.go]: Thanks for the reference!
Will update this part later.

File: shortcuts.go (hunk 6 above)
Comment (snkzt) [reply shortcuts.go]: Updated to print the error!

Comment (kinbiko) [reply shortcuts.go]: That's worse since it doesn't stop the execution of your program. 😅
**return** the error instead.

Comment (snkzt) [reply shortcuts.go]: Updated in the change [cb47f77](https://github.com/snkzt/shortcut-peeper/pull/2/commits/cb47f7760c67ccf8b8f138393a4d9a65ed4b7f71)!

File: main.go (hunk 5 above)
Comment (snkzt) [reply main.go]: Updated in the change [cb47f77](https://github.com/snkzt/shortcut-peeper/pull/2/commits/cb47f7760c67ccf8b8f138393a4d9a65ed4b7f71)!

**Pull request merged**